man/rasc2.1
.Dd October 27, 2010
.Dt RASC2 1
.Os
.Sh NAME
.Nm rasc2
.Nd Radare Shellcode helper utility
.Sh SYNOPSIS
.Nm rasc2
.Op -cCxXtLV
.Op -A N
.Op -C N
.Op -E N
.Op -N N
.Op -a addr@off
.Op -s hexpair
.Op -i name
.Op -S file
.Op -l port
.Sh DESCRIPTION
rasc2 implements the most common practices to generate shellcodes with paddings with 0x41's ('A'), nops, breakpoints or enumerations (00 01 02 03 ..).
.Pp
rasc2 can serve a syscall proxy at a specified port with \-l [port]
.Pp
There's an hardcoded database of common used shellcodes that can be listed with \-L. Use \-i to specify which want you want to use.
.Pp
The output can be raw or in hexpairs (\-x) or C array (\-c).
.Pp
Use \-X to inline execute the specified. (useful for testing)
.Pp
To specify the 
.Bl -tag -width Fl
.It Fl A|-C|-E|-N Ar count
Specify the number of 'A's, breakpoints, enumerations and nops to be prefixed to the shellcode.
.It Fl a Ar addr@off
Set the return address at a specified offset inside the resulting shellcode
.It Fl s Ar 'hexpair'
Set the shellcode in hexpair format. For example (\-s '01 02 03')
.It Fl l Ar port
Listen a syscall proxy server at a specified port.
.It Fl L
List all hardcoded shellcodes
.It Fl c
Show output in C array format
.It Fl x
Show output in hexpairs format
.It Fl X
Execute resulting shellcode
.It Fl t
Performs some tests on the current platform
.It Fl h
Show usage help message.
.El
.Sh ENVIRONMENT
Some shellcodes can be modified using the environment variables CMD, HOST and PORT. See \-L for more information.
.Sh SEE ALSO
.Pp
.Xr rahash2(1) ,
.Xr rafind(1) ,
.Xr radare2(1) ,
.Xr radiff2(1) ,
.Xr ranal2(1) ,
.Xr rabin2(1) ,
.Xr rasm2(1) ,
.Xr rax2(1) ,
.Xr ragg2(1) ,
.Xr rarun2(1) ,
.Sh AUTHORS
.Pp
pancake <pancake@nopcode.org>,
nibble <nibble@develsec.org>


ChangeLog
-=== RELEASE 0.8 ===-
changeset:   1625:0aaad74ad9fc
bookmark:    master
tag:         tip
user:        pancake
date:        Fri Jul 15 18:13:00 2011 +0200
files:       configure configure.acr libr/core/core.c libr/io/undo.c r2-bindings/python/r_cons.i.experimental
description:
* Bump release 0.8 - Codename thecakeisapie
* Fix uninitialized memory issue in r_block_resize()
  - Thanks vext01 for reporting


changeset:   1624:1315774824fb
parent:      1622:fc2643c36e30
parent:      1623:95f09764122e
user:        pancake
date:        Fri Jul 15 18:07:56 2011 +0200
description:
* Merge heads


changeset:   1623:95f09764122e
parent:      1620:236b81674ab5
user:        earada
date:        Fri Jul 15 14:17:36 2011 +0200
files:       libr/bin/p/bin_elf.c
description:
* Fix virtual address in elf without setcions


changeset:   1622:fc2643c36e30
user:        pancake
date:        Fri Jul 15 12:57:30 2011 +0200
files:       mk/gitpush.sh
description:
* Fix some typos in gitpush script


changeset:   1621:b6a6d3012998
user:        pancake
date:        Fri Jul 15 12:48:46 2011 +0200
files:       Makefile mk/gitpush.sh
description:
* Added gitpush utility for github mirroring


changeset:   1620:236b81674ab5
user:        pancake@pair
date:        Fri Jul 15 02:21:40 2011 +0200
files:       configure configure.acr
description:
* Bump acr to 0.8.6 with shellscript endian check
  - New flags --with-{little|big}-endian


changeset:   1619:d49dd6f2e789
user:        pancake@pair
date:        Fri Jul 15 01:50:42 2011 +0200
files:       binr/Makefile doc/maemo libr/Makefile maemo/Makefile maemo/README maemo/deb_hand.mak maemo/description.txt
description:
* Added support for new platform Maemo6-Harmattan
  - Linux/arm with debugger support
  - Debian package generation tools (no need for dpkg)
  - Compilation instructions at doc/maemo
* Fix install in binr and libr


changeset:   1618:dd1c50f121bf
user:        pancake@pair
date:        Thu Jul 14 22:14:11 2011 +0200
files:       configure configure.acr
description:
* Handle -lssl and pkg-config openssl
  - Should fix NetBSD build


changeset:   1617:4c6e8c50e7cf
user:        earada
date:        Thu Jul 14 16:17:39 2011 +0200
files:       libr/debug/p/debug_native.c libr/util/sys.c
description:
* Fix FreeBSD build


changeset:   1616:fd36edc76e42
user:        pancake
date:        Wed Jul 13 19:37:28 2011 +0200
files:       test/t/asm
description:
* Fix asm test


changeset:   1615:1ad69f14f1db
user:        pancake
date:        Wed Jul 13 19:28:53 2011 +0200
files:       test/tests.sh
description:
* Fix test suite


changeset:   1614:6c552be332cb
user:        pancake
date:        Wed Jul 13 19:22:38 2011 +0200
files:       binr/radare2/Makefile binr/radiff2/Makefile binr/rafind2/Makefile config-user.mk.acr configure configure.acr libr/debug/t/Makefile libr/hash/t/Makefile libr/include/r_userconf.h.acr libr/io/p/debug.mk libr/io/p/gdb.mk libr/io/p/haret.mk libr/io/t/Makefile libr/socket/Makefile libr/socket/t/Makefile
description:
* Use OpenSSL (cflags/ldflags) from pkg-config


changeset:   1613:8ecc34f29bc4
user:        pancake
date:        Wed Jul 13 18:55:16 2011 +0200
files:       doc/windows libr/asm/arch/msil/demsil.c libr/asm/p/asm_dalvik.c libr/bin/format/mach0/dyldcache.c libr/config.mk.tail libr/core/vmenus.c libr/include/list.h libr/include/r_types_base.h libr/io/p/io_w32dbg.c libr/socket/t/test.c libr/util/cache.c
description:
* Fix build and lot of warnings for w64


changeset:   1612:1552f58ae98c
user:        earada
date:        Wed Jul 13 17:41:26 2011 +0200
files:       binr/radare2/radare2.c libr/core/cmd.c libr/core/config.c libr/core/file.c libr/include/r_core.h
description:
* Use native api to calculate md5/sha1 hashes
- Fix bug 29


changeset:   1611:3dd0b00836fe
user:        pancake
date:        Wed Jul 13 13:59:55 2011 +0200
files:       binr/rax2/rax2.c doc/fortunes libr/fs/Makefile libr/fs/p/grub/Makefile libr/include/r_io.h libr/io/p/io_w32dbg.c
description:
* Fix w32 and w64 build (r_fs/grub issue)
* Dont quit rax2 when reading from stdin until 'q' or ^D
* Minor fix in fortunes


changeset:   1610:2637b35a1b31
user:        pancake
date:        Wed Jul 13 12:21:56 2011 +0200
files:       libr/core/cmd.c
description:
* Fix double pdf (thanks vext01)


changeset:   1609:5b507dbd2c5e
user:        pancake
date:        Wed Jul 13 11:30:25 2011 +0200
files:       Makefile
description:
* Remove manpages in purge
* Fix make symstall


changeset:   1608:980201d7e995
user:        pancake@pair
date:        Wed Jul 13 10:14:30 2011 +0200
files:       binr/binr.mk binr/radare2/Makefile binr/radiff2/Makefile binr/rafind2/Makefile doc/static libr/debug/p/libgdbwrap/gdbwrapper.c libr/debug/p/libgdbwrap/include/gdbwrapper-internals.h libr/debug/p/libgdbwrap/include/gdbwrapper.h libr/debug/t/Makefile libr/hash/t/Makefile libr/io/Makefile libr/io/p/debug.mk libr/io/p/gdb.mk libr/io/p/haret.mk libr/io/p/io_gdb.c libr/io/p/mach.mk libr/io/p/malloc.mk libr/io/p/shm.mk libr/io/t/Makefile libr/socket/t/Makefile mk/gcc.mk
description:
* Fix static build (--without-pic --with-nonpic)


changeset:   1607:3a06619b5226
user:        pancake@pair
date:        Tue Jul 12 23:22:33 2011 +0200
files:       libr/core/asm.c libr/core/cmd.c
description:
* Fix pi and /c commands


changeset:   1606:e220fccf6bc4
user:        earada
date:        Tue Jul 12 23:06:06 2011 +0200
files:       libr/core/cmd.c
description:
* Fix some warnings


changeset:   1605:d15cbb5291a3
user:        pancake
date:        Tue Jul 12 19:42:20 2011 +0200
files:       doc/windows libr/io/p/io_w32dbg.c mk/x86_64-w64-mingw32-gcc.mk
description:
* Fix build for ming-w64


changeset:   1604:026708c1cb88
user:        earada
date:        Tue Jul 12 17:26:56 2011 +0200
files:       libr/core/vmenus.c libr/print/format.c
description:
* Enhacements in keybindings used in visual mounts menus
* Implement missing 'pf p' pointer format


changeset:   1603:71f82c98175a
user:        pancake@pairch

binr/rasc2/rasc2.c
/* radare - LGPL - Copyright 2007-2011 pancake <@nopcode.org> */

#include "rasc.h"
#include "r_types.h"
#include "r_userconf.h"
#include <stdio.h>
#include <string.h>
#include <getopt.h>
#include <stdlib.h>
#include <sys/types.h>
#include <unistd.h>
#include <r_util.h>

#if __UNIX__
#include <sys/mman.h>
#include <sys/socket.h>
#include <netinet/in.h>
#endif

#if __WINDOWS__
#include <windows.h>
#endif

#define BLOCK 4096
static const char *ofile = NULL;
static const char *encoder = NULL;
static int scidx = -1;
static int hexa_print = 0;
static ut32 off = 0, addr = 0;
static ut8 shellcode[BLOCK];
static ut8 output[BLOCK];

/* sizes */
static int A=0, N=0;
static int C=0, E=0;
static int scsize = 0;
#define SCSIZE N+A+C+E+scsize

static int show_helpline() {
	printf( "Usage: rasc2 [-cCexXtLV] [-l port] [-a addr@off] [-[A|N|C|E] N]\n"
		"             [-s hexpair] [-i name] [-S file] [-h]\n");
	return 0;
}

static int show_help() {
	show_helpline ();
	printf (
	"  -l [port]    starts a syscall proxy server\n"
	"  -A [n]       prefix shellcode with N A's (0x41)\n"
	"  -N [n]       prefix shellcode with N nops (0x90)\n"
	"  -C [n]       suffix shellcode with N traps\n"
	"  -E [n]       prefix with enumeration 01 02 03..\n"
	"  -a addr@off  set the return address at a specified offset\n"
	"  -s 'sc'      set shellcode in hexpairs (cc cd 80)\n"
	"  -S 'file'    load shellcode from file\n"
	"  -i 'scdb'    hardcoded shellcode (-L to list)\n"
	"  -L           list hardcoded shellcodes\n"
	"  -c           output in C format\n"
	"  -e           output in escapped string\n"
	"  -x           output in hexpairs format\n"
	"  -O [encoder] select output encoder (fmi: -O help)\n"
	"  -o [file]    select output file\n"
	"  -X           execute shellcode\n"
	"  -t           test current platform\n"
	"  -V           show version information\n"
	" Environment variables to compile shellcodes:\n"
	"  CMD          Command to execute on execves\n"
	"  HOST         Host to connect\n"
	"  PORT         Port to listen or connect\n");
//#warning TODO: prefix shellcode with setuid() fun
	//printf("  -p          attach \n");
	//printf("  -P          push file and remote execute\n");
	//printf("  -u          use UDP\n");
	return 0;
}

int encode (const char *encoder, ut8 *dst, int dstlen, ut8 *src, int srclen) {
	int xordeclen, i;
	if (!strcmp (encoder, "xor")) {
		//ut8 key = 33;
		// Find valid xor key
		// length is key here
		const ut8 *xordec = (const ut8*)
			// TODO: setup ecx here
			"\xe8\xff\xff\xff\xff" // call $$+4
			"\xc1" // ffc1 = inc ecx
			"\x5e" // pop esi
			"\x30\x4c\x0e\x07" // xor [esi+ecx+7], cl
			"\xe2\xfa"; // loop xoresi
		xordeclen = strlen ((const char *)xordec);
		if (srclen+xordeclen>=dstlen) {
			eprintf ("encode: too long");
			return 0;
		}
		memcpy (dst, xordec, xordeclen);
		for (i=0;i<srclen; i++) {
			dst[xordeclen+i] = src[i] ^ i; // XXX
		}
		memcpy (dst+xordeclen, src, srclen);
		return srclen + xordeclen;
	} else {
		eprintf ("Encoders: xor\n");
		exit (0);
	}
	return 0;
}

char *filetostr(char *file) {
        FILE *fd = fopen (file,"r");
        char *buf;
        int i, size = BLOCK;

        if (fd == NULL)
                return NULL;

        buf = (char *)malloc (size);
        buf[0]='\0';
        for (i=0; !feof (fd); i++) {
                if (i==size) {
                        size = size + BLOCK;
                        buf = realloc (buf, size);
                }
                fread (buf+i, 1, 1, fd);
        }
        fclose (fd);
        if (buf[0]=='\0') {
                free (buf);
                return NULL;
        }
        return buf;
}

int otf_patch() {
	char *ptr;
	/* on the fly patching */
	if (scidx != -1) {
		if (shellcodes[scidx].cmd) {
			ptr = getenv ("CMD");
			if (ptr) {
				strcpy((char*) (shellcode+shellcodes[scidx].cmd), ptr);
				shellcode[shellcodes[scidx].cmd+strlen(ptr)]='\0';
				if (strlen(ptr)>7)
					scsize+=strlen(ptr)-7;
			}
		}
		if (shellcodes[scidx].host) {
			ptr = getenv ("HOST");
			if (ptr) {
				int x,y,z,w;
				sscanf (ptr,"%d.%d.%d.%d", &x,&y,&z,&w);
				shellcode[shellcodes[scidx].host+3]=x;
				shellcode[shellcodes[scidx].host+2]=y;
				shellcode[shellcodes[scidx].host+1]=z;
				shellcode[shellcodes[scidx].host+0]=w;
			}
		}
		if (shellcodes[scidx].port) {
			ptr = getenv ("PORT");
			if (ptr) {
				unsigned short port = atoi(ptr);
				memcpy (shellcode+shellcodes[scidx].port, &port, 2);
			}
		}
	}
	/* patch return address */
	if (addr != 0) {
		/* TODO: swapping endian for addr (-e) */
		ut8 *foo = (ut8 *)&addr;
		if (off<0) off = 0;
		if (off>SCSIZE) off = SCSIZE-4;
		output[off+0] = foo[0];
		output[off+1] = foo[1];
		output[off+2] = foo[2];
		output[off+3] = foo[3];
	}
	return 0;
}

int print_shellcode() {
	int j=0, i=0;

	if (!(SCSIZE)) {
		printf ("No shellcode defined\n");
		return 1;
	}

	if (SCSIZE>=BLOCK) {
		printf ("Dont overflow me\n");
		return 1;
	}

	/* prepare output buffer */
	for (i=0;i<A;i++)
		output[i] = 'A';
	if (N%2) {
		for (i=0; i<N; i++)
			output[i+A] = '\x90';
	} else {
		for (i=0; i<N; i+=2) {
			output[i+A]   = '\x40'; // inc eax
			output[i+A+1] = '\x48'; // dec eax
		}
	}
	for(i=0,j='A'; i<E; i++,j++) {
		if (j=='\n'||j=='\r')
			j++;
		output[i*4+A+N] = (ut8)(j%256);
		output[i*4+A+N+1] = (ut8)(j%256);
		output[i*4+A+N+2] = (ut8)(j%256);
		output[i*4+A+N+3] = (ut8)(j%256);
	}
	/* patch addr and env */
	otf_patch ();

	if (encoder) {
		ut8 blob[BLOCK];
		scsize = encode (encoder, blob, sizeof (blob), shellcode, scsize);
		memcpy (output+A+N+E, blob, scsize);
	} else memcpy (output+A+N+E, shellcode, scsize);
	for (i=0; i<C; i++)
		output[i+A+E+N+scsize] = '\xCC';

	if (ofile) {
		int fd;
		unlink (ofile);
		fd = open (ofile, O_RDWR | O_CREAT, 0755);
		dup2 (fd, 1);
	}
	switch (hexa_print) {
	case 0: // raw
		write (1, output, SCSIZE);
		break;
	case 1: // hexpairs
		for (i=0; i<SCSIZE; i++)
			printf ("%02x", output[i]);
		printf ("\n");
		break;
	case 2: // C
		printf ("ut8 shellcode[] = {  ");
		j = 0;
		for (i=0; i<SCSIZE; i++) {
			if (!(i%12)) printf ("\n  ");
			printf ("0x%02x", output[i]);
			if (i+1!=SCSIZE+scsize)
				printf (", ");
		}
		printf ("\n};\n");
		break;
	case 3:
		if (scsize == 0) {
			printf("No shellcode defined\n");
			return 1;
		} else {
			ut8 *ptr = malloc (4096);
			void (*cb)() = (void *)&shellcode;
			memcpy (ptr, shellcode, SCSIZE);
			r_mem_protect (ptr, 4096, "rx");
			r_mem_protect (ptr, 4096, "rwx"); // try, ignore if fail
			cb = (void*)ptr;
			cb ();
			free (ptr);
		}
		break;
	case 4:
		printf ("\"");
		j = 0;
		for (i=0;i<SCSIZE;i++)
			printf ("\\x%02x", output[i]);
		printf ("\"\n");
		break;
	}
	fflush (stdout);
	if (ofile)
		close (1);
	return 0;
}

int hex2int (ut8 *val, ut8 c) {
        if ('0' <= c && c <= '9')      *val = (ut8)(*val) * 16 + ( c - '0');
        else if (c >= 'A' && c <= 'F') *val = (ut8)(*val) * 16 + ( c - 'A' + 10);
        else if (c >= 'a' && c <= 'f') *val = (ut8)(*val) * 16 + ( c - 'a' + 10);
        else return 0;
        return 1;
}

int hexpair2bin(const char *arg) { // (0A) => 10 || -1 (on error)
	ut8 *ptr;
	ut8 c = '\0';
	ut8 d = '\0';
	unsigned int j = 0;

	for (ptr = (ut8 *)arg; ; ptr++) {
		if (ptr[0]==' '||ptr[0]=='\t'||ptr[0]=='\n'||ptr[0]=='\r')
			continue;
		if (!IS_PRINTABLE(ptr[0]))
			continue;
		if (ptr[0]=='\0'||ptr[0]==' ' || j==2)
			break;
		d = c;
		if (hex2int(&c, ptr[0]) == 0) {
			fprintf(stderr, "Invalid hexa string at char '%c'.\n", ptr[0]);
			return -1;
		}
		c |= d;
		if (j++ == 0) c <<= 4;
	}

	return (int)c;
}

void cipher_memcpy(ut8 *dst, ut8 *src, int len) {
	int i, n;
	for (i=0; i<len; i++) {
		n = src[i] & 0xf;
		n = n==0xf?0:n+1;
		dst[i] = n;
		n = src[i] & 0xf0;
		n>>=4;
		n = n==0xf?0:n+1;
		dst[i] |= n<<4;
	}
}

int load_shellcode_from_me(char *str) {
	int i;
	for (i=0; shellcodes[i].name; i++) {
		if (!strcmp (shellcodes[i].name, str)) {
			//memcpy (shellcode, shellcodes[i].data, shellcodes[i].len);
			/* cipher shit */
			cipher_memcpy (shellcode, shellcodes[i].data, shellcodes[i].len);
			scsize = shellcodes[i].len;
			scidx = i;
//printf("Using %d bytes shellcode (%s) %02x %02x\n", shellcodes[i].len, shellcodes[i].desc,
//shellcodes[i].data[0], shellcodes[i].data[1], shellcodes[i].data[2]);
			return 1;
		}
	}
	return 0;
}

// XXX: return value is always 0??
int load_shellcode_from_string(char *str) {
	int i, j=1, ch, len;
	char input2[BLOCK];
	strncpy (input2, str, BLOCK-1);
	len = strlen (input2);
	input2[0] = '\0';
	for (i=0;i<len;i+=j) {
		if (str[i]==' '||str[i]=='\t'||str[i]=='\n'||str[i]=='\r')
			continue;
		ch = hexpair2bin (str+i);
		if (str[i+2]==' ')
			j = 3;
		else j = 2;
		if (ch == -1)
			break;
		shellcode[scsize++]=ch;
	}
	shellcode[scsize] = '\0';
	return 0;
}

static int file_type(char *str) {
	if (!strcmp(str,"-"))
		return 0; // stdin
	if (!strcmp(str+strlen(str)-2,".s"))
		return 1; // stdin
	return 2;
}

static int load_shellcode_from_file(char *str) {
	char buf[1024];
	char *ptr = NULL;

	eprintf ("TODO: This is r1-dependant.. ugly . must dump all disasm\n");
	str[1024]='\0';
	switch (file_type (str)) {
	case 0: // stdin
		fprintf (stderr, "TODO\n");
		break;
	case 1: // .s file (assembly
		sprintf (buf, "gcc -nostdlib -o .x %s", str);
		system (buf);
		system ("rsc syms-dump .x | grep _start | cut -d : -f 2 | tee .y");
		unlink (".x");
		ptr = filetostr (".y");
		unlink (".y");
		if (ptr) {
			load_shellcode_from_string(ptr);
			free(ptr);
		}
		break;
	default:
		eprintf ("File format not supported\n");
		exit (1);
	}

	return 0;
}

int main(int argc, char **argv) {
	int c, listen = 0;

	if (argc<2)
		return show_helpline ();

	while ((c = getopt (argc, argv, "a:VcC:ts:S:i:Ll:uhN:A:XxE:eo:O:")) != -1) {
		switch (c) {
		case 'o':
			// output file
			ofile = optarg;
			break;
		case 'O':
			// output encoder
			eprintf ("TODO: no encoders implemented yet\n");
			encoder = optarg;
			break;
		case 't':
			return test ();
		case 'x':
			// dump shellcode in hexa
			hexa_print = 1;
			break;
		case 'X':
			// execute
			hexa_print = 3;
			break;
		case 'C':
			C = atoi (optarg);
			break;
		case 'E':
			E = atoi (optarg);
			break;
			// dump shellcode in C
		case 'e':
			hexa_print = 4;
			break;
		case 'c':
			hexa_print = 2;
			break;
		case 'a':
			sscanf (optarg, "%x@%x", (int*) &addr, (int*) &off);
			if (!addr||!off)
				sscanf (optarg, "0%x@%x", (int*) &addr, (int*) &off);

			if (!addr||!off) {
				printf ("Invalid argument for -a\n");
				return 1;
			}
			break;
		case 'A':
			A = atoi (optarg);
			break;
		case 's':
			load_shellcode_from_string (optarg);
			break;
		case 'S':
			load_shellcode_from_file (optarg);
			break;
		case 'i':
			if (!load_shellcode_from_me (optarg)) {
				printf ("Cannot find shellcode '%s'\n", optarg);
				return 1;
			}
			break;
		case 'N':
			N = atoi (optarg);
			break;
		case 'V':
			printf ("rasc2 "R2_VERSION"\n");
			return 0;
		case 'p':
			// prefix the contents of this file
			break;
		case 'h':
			return show_help ();
		case 'l':
			listen = atoi (optarg);
			break;
		case 'u':
			printf ("TODO: UDP support\n");
			break;
		case 'L':
			for (c=0;shellcodes[c].name;c++) {
				printf ("%-20s  %3d   %s\n",
					shellcodes[c].name,
					shellcodes[c].len,
					shellcodes[c].desc);
			}
			return 0;
		}
	}
	print_shellcode ();
	return 0;
}
